opal/corelib/string.rb

```diff
@@ -175,7 +191,7 @@ def =~(other)
 
   def [](index, length = undefined)
     %x{
-      var size = self.length, exclude, range;
+      let str = [...self], size = str.length, exclude, range;
```

Comment:
My idea for some optimization:

A UTF-16 character of index N won't happen later than in N*2th 16bit word, so we could just expand a slice. And calls like `a[0]`, `a[1]` will be the most common, then `a[a.length-1]` will be the second most common, for which we could think of doing a similar but in reverse.

Reply:
Hm, for the reverse, length will be a harder to procure though...

Reply:
We should not need length with a negative index, we can just walk from the back.